### PR TITLE
add support for shared transitions

### PR DESCRIPTION
**Pull request by t1amat9409**
Added support for https://beta.docs.nativescript.org/guide/shared-element-transitions, exposes `$navigate` (can be used to apply the shared Elements Transitions) and `$showModal` as inspired by `nativescript-vue`.

Usage:
```TS
const HomePage = () => {
 const { $navigate } = useNativeScreensContext();

 // call the $navigate method
$navigate({
      screen: ShareTransionScreenTest,
      props: {
        userId: 1,
      },
      transition: SharedTransition.custom(new PageTransition(), {
        pageStart: {
          x: 0,
          y: 300,
          opacity: 0,
          height: 0,
        },
        pageEnd: {
          x: 0,
          y: 100,
          duration: 256,
          opacity: 1,
        },
        pageReturn: {
          y: 600,
          duration: 250,
          useStartOpacity: true,
        }
      }),
      pageProps: {
        stackPresentation: 'push',
        actionBarHidden: true,
        gestureEnabled: true,
      }
    })
}
```

### CODE REVIEW

**Comment by t1amat9409** [comment]
@shirakaba please take a look at this PR and let me know what you think

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 1]
```diff
@@ -178,81 +179,67 @@ export interface ScreenStackHeaderConfigProps extends FlexboxLayoutAttributes {
...
+          ENABLE_SCREENS && !active && !onScreen ?
+            {
+              visibility: 'collapse' // Because `display: 'none'` doesn't exist in NativeScript
+            } :
+            {}
+        ),
+      }}
+      {...rest}
+    />
+  );
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:242]: Thanks for converting this to a hooks-based component. May improve HMR.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 2]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+  $navigate<T>(params: GoToScreenParams<T>): void
+  $goBack(): void
+  $closeModal(): void
+}
+
+const NativeScreensContext = React.createContext({} as NativeScreensContextData)
+
+export const NativeScreensProvider: React.FC<FrameProps> = ({ children, ...rest }) => {
+  const navRef = React.useRef<NSVElement<Frame>>()
+  function createView<T>(params: GoToScreenParams<T>, callback = (page: Page) => { }, modal?: boolean) {
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:450]: Could this be hoisted to the outer scope to avoid re-creating the function on each render? I don't think it has any dependencies on anything inside NativeScreensProvider.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 3]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+    const { screen: Screen, props } = params
+    if (!navRef.current) return
+    const ref = React.createRef<NSVElement<ContentView>>()
+    const container = new NSVElement<Page>('page')
+    if (modal) container.nativeView.background = 'transparent'
+    const key = `${Date.now()}-screen`
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:456]: May be better to calculate a [cheap UUID](https://github.com/lukeed/uuid) than rely on the date here.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 4]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+}
+
+const NativeScreensContext = React.createContext({} as NativeScreensContextData)
+
+export const NativeScreensProvider: React.FC<FrameProps> = ({ children, ...rest }) => {
+  const navRef = React.useRef<NSVElement<Frame>>()
+  function createView<T>(params: GoToScreenParams<T>, callback = (page: Page) => { }, modal?: boolean) {
+    const { screen: Screen, props } = params
+    if (!navRef.current) return
+    const ref = React.createRef<NSVElement<ContentView>>()
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:453]: Is this being used anywhere?

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 5]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+  pageProps?: ScreenProps
+}
+interface NativeScreensContextData {
+  $showModal<T>(params: GoToScreenParams<T>): void
+  $navigate<T>(params: GoToScreenParams<T>): void
+  $goBack(): void
+  $closeModal(): void
+}
+
+const NativeScreensContext = React.createContext({} as NativeScreensContextData)
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:446]: I ran into this recently – I think we can handle the initialisation more elegantly than passing an empty object with a type assertion.

Instead, we can let the context also accept `undefined` and just pass in that. To avoid users having to handle `undefined` downstream, we'd then wrap the context so that users import `useNativeScreensContext` rather than directly importing `NativeScreensContext`. This is a Kent C Dodds trick: https://kentcdodds.com/blog/how-to-use-react-context-effectively

Example:

image

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 6]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+              container.nativeView[key] = params.pageProps[key]
+            }
+          })
+        }
+        callback(container.nativeView)
+      }, key, true)
+  }
+  function $showModal<T>(params: GoToScreenParams<T>, fullScreen?: boolean) {
+    createView(params, (page) => {
+      navRef.current.nativeView.showModal(page, {
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:478]: Best to handle the case of `navRef.current` being falsy, with a warning to the user.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 7]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+        </NativeScreensProvider>
+      ),
+      container, () => {
+        if (params.pageProps) { 
+          Object.keys(params.pageProps).forEach((key) => {
+            if (container.nativeView[key]) {
+              container.nativeView[key] = params.pageProps[key]
+            }
+          })
+        }
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:472]: I'm a little nervous about this because it's imperative props-handling. I don't see the whole picture, but is there a way that the user might navigate back to the same page and React might re-use the same component (and thus fail to reset all the previously-set pageProps back to normal on next render?)

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 8]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+export const NativeScreensProvider: React.FC<FrameProps> = ({ children, ...rest }) => {
+  const navRef = React.useRef<NSVElement<Frame>>()
+  function createView<T>(params: GoToScreenParams<T>, callback = (page: Page) => { }, modal?: boolean) {
+    const { screen: Screen, props } = params
+    if (!navRef.current) return
+    const ref = React.createRef<NSVElement<ContentView>>()
+    const container = new NSVElement<Page>('page')
+    if (modal) container.nativeView.background = 'transparent'
+    const key = `${Date.now()}-screen`
+    render(
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:457]: I'm a little nervous about having to use React NativeScript's render function (although yes, I do this for ListView for synchronous rendering – which indeed I am also nervous about). I wonder if there's a way to do this declaratively instead, like with a root-level container that users have to opt into. But I haven't thought too hard to figure out the limitations.

File: react-nativescript-navigation/src/native-stack/views/FrameNavigatorView.tsx [hunk 9]
```diff
@@ -5,8 +5,8 @@
 import * as React from 'react';
 // import { View, StyleSheet, Platform } from 'react-native';
 import {
-  ScreenStack,
-  Screen as ScreenComponent,
+  NativeScreensProvider as ScreenStack,
+  NativeScreen as ScreenComponent,
```
Comment (shirakaba) [inline react-nativescript-navigation/src/native-stack/views/FrameNavigatorView.tsx:9]: Just a note to myself to look at the consequences of changing ScreenStack from strictly `"frame"` to `NativeScreensProvider`. I'm nervous about regressing other features to support shared transitions.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 10]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+        transition: params.transition,
+        backstackVisible: true,
+        context: params.props,
+        animated: true,
+      })
+    })
+  }
+
+  function $goBack() {
+    if (Frame.topmost().canGoBack()) Frame.topmost().goBack()
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:504]: For one thing, I think we'll want a truthy check:
```suggestion
    if(Frame.topmost()?.canGoBack()){
      Frame.topmost().goBack();
    }
```
For another, I think ideally we should get a ref to the frame itself rather than relying on `Frame.topmost()`, as things may get complicated with nested frames.

`useCallback` would be worth doing, too.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 11]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+  }
+
+  function $closeModal() {
+    Frame.topmost().closeModal()
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:508]:
```suggestion
    Frame.topmost()?.closeModal();
```
(And same as above, ideally we should get a ref to the frame itself, and use `useCallback`).

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 12]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+        context: params.props,
+        closeCallback: (_args: any) => {
+
+        }
+      });
+    })
+  }
+  function $navigate<T>(params: GoToScreenParams<T>) {
+    createView(params, (page) => {
+      navRef.current.nativeView.navigate({
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:491]: Best to handle the case of `navRef.current` being falsy, with a warning to the user.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 13]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+          Object.keys(params.pageProps).forEach((key) => {
+            if (container.nativeView[key]) {
+              container.nativeView[key] = params.pageProps[key]
+            }
+          })
+        }
+        callback(container.nativeView)
+      }, key, true)
+  }
+  function $showModal<T>(params: GoToScreenParams<T>, fullScreen?: boolean) {
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:476]: While we can't hoist this (as it depends on `navRef`), how about we use `useCallback` to memoise it.

File: react-nativescript-navigation/src/react-nativescript-screens/screens.tsx [hunk 14]
```diff
@@ -442,3 +429,96 @@ export const NativeScreenContainer: React.ElementType<JSX.IntrinsicElements["fle
...
+        transparent: true,
+        fullscreen: fullScreen,
+        context: params.props,
+        closeCallback: (_args: any) => {
+
+        }
+      });
+    })
+  }
+  function $navigate<T>(params: GoToScreenParams<T>) {
```
Comment (shirakaba) [inline react-nativescript-navigation/src/react-nativescript-screens/screens.tsx:489]: While we can't hoist this (as it depends on `navRef`), how about we use `useCallback` to memoise it.

**Review by shirakaba — COMMENTED** [review]
Thanks for the PR, it's a cool feature that I'd have been out of my depth to implement myself!

That said, it is also a complex feature and I'm not fully confident about how it all works. I've had a look over the code to see if it's an easy win, but upon inspection, I'm not too sure about some parts.

As I'm not actively maintaining React NativeScript beyond life support, I'm afraid I don't think I have the time to test this out and check for edge cases. I wonder if it may be possible to implement this as an external add-on, as I see it largely wraps existing things.

**Comment by t1amat9409** [comment]
It's been 3 years already, not sure if this still works😂😂

**Comment by shirakaba** [comment]
Yeah, I also haven't touched React NativeScript during that whole time, so I'd say I'd recommend just maintaining your own fork of React NativeScript Navigation if you still need this feature 😅

For the sake of sustainability and maintainability, I intend for the the future of React NativeScript to be "NativeScript in React Native". I've been pushing on that dream for ages. For the last few years, I was pushing React Native to adopt Node-API so that we could use [NativeScript Node-API](https://github.com/NativeScript/napi-ios). But just last month, [NativeScript JSI](https://www.npmjs.com/package/@nativescript/react-native) was released (currently in preview, so not stable). So in theory, you'd just install `@nativescript/react-native` into a React Native app and then you'd be able to access NativeScript from inside a React Native app. I haven't tried it yet, though.

Thanks for being one of React NativeScript's earliest users! 🙇‍♂️
